modules/pubmaticBidAdapter.js

```diff
@@ -0,0 +1,295 @@
+  var conf = {};
+  conf.pageURL = utils.getTopWindowUrl();
+  conf.refURL = utils.getTopWindowReferrer();
+  return conf;
+}
+
+function _handleCustomParams(params, conf) {
+  // istanbul ignore else
+  if (!conf.kadpageurl) {
+    conf.kadpageurl = conf.pageURL;
+  }
+
+  var key, value, entry;
+  for (key in CUSTOM_PARAMS) {
+    // istanbul ignore else
+    if (CUSTOM_PARAMS.hasOwnProperty(key)) {
+      value = params[key];
+      // istanbul ignore else
+      if (value) {
+        entry = CUSTOM_PARAMS[key];
+
+        if (typeof entry === 'object') {
+          // will be used in future when we want to process a custom param before using
+          // 'keyname': {f: function() {}}
+          value = entry.f(value, conf);
+        }
+
+        if (utils.isStr(value)) {
+          conf[key] = value;
+        } else {
+          utils.logWarn('PubMatic: Ignoring param : ' + key + ' with value : ' + CUSTOM_PARAMS[key] + ', expects string-value, found ' + typeof value);
+        }
+      }
+    }
+  }
+  return conf;
+}
+
+function _createOrtbTemplate(conf) {
+  return {
+    id: '' + new Date().getTime(),
+    at: AUCTION_TYPE,
+    cur: [CURRENCY],
+    imp: [],
+    site: {
+      page: conf.pageURL,
+      ref: conf.refURL,
+      publisher: {}
+    },
+    device: {
+      ua: navigator.userAgent,
+      js: 1,
+      dnt: (navigator.doNotTrack == 'yes' || navigator.doNotTrack == '1' || navigator.msDoNotTrack == '1') ? 1 : 0,
+      h: screen.height,
+      w: screen.width,
+      language: navigator.language
+    },
+    user: {},
+    ext: {}
+  };
+}
+
+function _createImpressionObject(bid, conf) {
+  return {
+    id: bid.bidId,
+    tagid: bid.params.adUnit,
+    bidfloor: _parseSlotParam('kadfloor', bid.params.kadfloor),
+    secure: window.location.protocol === 'https:' ? 1 : 0,
+    banner: {
+      pos: 0,
+      w: bid.params.width,
+      h: bid.params.height,
+      topframe: utils.inIframe() ? 0 : 1,
+    },
+    ext: {
+      pmZoneId: _parseSlotParam('pmzoneid', bid.params.pmzoneid)
+    }
+  };
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+
+  /**
+  * Determines whether or not the given bid request is valid. Valid bid request must have placementId and hbid
+  *
+  * @param {BidRequest} bid The bid params to validate.
+  * @return boolean True if this is a valid bid, and false otherwise.
+  */
+  isBidRequestValid: bid => {
+    if (bid && bid.params) {
+      if (!utils.isStr(bid.params.publisherId)) {
+        utils.logWarn('PubMatic Error: publisherId is mandatory and cannot be numeric. Call to OpenBid will not be sent.');
+        return false;
+      }
+      if (!utils.isStr(bid.params.adSlot)) {
+        utils.logWarn('PubMatic: adSlotId is mandatory and cannot be numeric. Call to OpenBid will not be sent.');
+        return false;
+      }
+      return true;
+    }
+    return false;
+    // return !!(bid && bid.params && utils.isStr(bid.params.publisherId) && utils.isStr(bid.params.adSlot));
```

Comment:
Can we remove this comment if the earlier code is the planned/definite approach to perform the validation check?